File: js/core/src/background-action.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { JSONSchema7 } from 'json-schema';
import * as z from 'zod';
import { Action, ActionMetadata, defineAction, Middleware } from './action.js';
import { ActionContext } from './context.js';
import { GenkitError } from './error.js';
import { ActionType, Registry } from './registry.js';
import { toJsonSchema } from './schema.js';

/**
 * Zod schema of an opration representing a background task.
 */
export const OperationSchema = z.object({
  action: z.string().optional(),
  id: z.string(),
  done: z.boolean().optional(),
  output: z.any().optional(),
  error: z.object({ message: z.string() }).passthrough().optional(),
  metadata: z.record(z.string(), z.any()).optional(),
});

/**
 * Background operation.
 */
export interface Operation<O = any> {
  action?: string;
  id: string;
  done?: boolean;
  output?: O;
  error?: { message: string; [key: string]: unknown };
  metadata?: Record<string, any>;
}

/**
 * Background action. Unlike regular action, background action can run for a long time in the background.
 * The returned operation can used to check the status of the background operation and retrieve the response.
 */
export interface BackgroundAction<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
  RunOptions extends BackgroundActionRunOptions = BackgroundActionRunOptions,
> {
  __action: ActionMetadata<I, O>;
  readonly supportsCancel: boolean;

  start(
    input?: z.infer<I>,
    options?: RunOptions
  ): Promise<Operation<z.infer<O>>>;

  check(operation: Operation<z.infer<O>>): Promise<Operation<z.infer<O>>>;

  cancel(operation: Operation<z.infer<O>>): Promise<Operation<z.infer<O>>>;
}

export async function lookupBackgroundAction<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
>(
  registry: Registry,
  key: string
): Promise<BackgroundAction<I, O> | undefined> {
  const root: Action<I, typeof OperationSchema> = await registry.lookupAction<
    I,
    typeof OperationSchema,
    Action<I, typeof OperationSchema>
  >(key);
  if (!root) return undefined;
  const actionName = key.substring(key.indexOf('/', 1) + 1);
  return new BackgroundActionImpl(
    root,
    await registry.lookupAction<
      typeof OperationSchema,
      typeof OperationSchema,
      Action<typeof OperationSchema, typeof OperationSchema>
    >(`/check-operation/${actionName}/check`),
    await registry.lookupAction<
      typeof OperationSchema,
      typeof OperationSchema,
      Action<typeof OperationSchema, typeof OperationSchema>
    >(`/cancel-operation/${actionName}/cancel`)
  );
}

class BackgroundActionImpl<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
  RunOptions extends BackgroundActionRunOptions = BackgroundActionRunOptions,
> implements BackgroundAction<I, O, RunOptions>
{
  __action: ActionMetadata<I, O>;

  readonly startAction: Action<I, typeof OperationSchema>;
  readonly checkAction: Action<typeof OperationSchema, typeof OperationSchema>;
  readonly cancelAction?: Action<
    typeof OperationSchema,
    typeof OperationSchema
  >;

  constructor(
    startAction: Action<I, typeof OperationSchema>,
    checkAction: Action<typeof OperationSchema, typeof OperationSchema>,
    cancelAction:
      | Action<typeof OperationSchema, typeof OperationSchema>
      | undefined
  ) {
    this.__action = {
      name: startAction.__action.name,
      description: startAction.__action.description,
      inputSchema: startAction.__action.inputSchema,
      inputJsonSchema: startAction.__action.inputJsonSchema,
      metadata: startAction.__action.metadata,
      actionType: startAction.__action.actionType,
    };
    this.startAction = startAction;
    this.checkAction = checkAction;
    this.cancelAction = cancelAction;
  }

  async start(
    input?: z.infer<I>,
    options?: RunOptions
  ): Promise<Operation<z.infer<O>>> {
    return await this.startAction(input, options);
  }

  async check(
    operation: Operation<z.infer<O>>
  ): Promise<Operation<z.infer<O>>> {
    return await this.checkAction(operation);
  }

  get supportsCancel(): boolean {
    return !!this.cancelAction;
  }

  async cancel(
    operation: Operation<z.infer<O>>
  ): Promise<Operation<z.infer<O>>> {
    if (!this.cancelAction) {
      return operation;
    }
    return await this.cancelAction(operation);
  }
}

/**
 * Options (side channel) data to pass to the model.
 */
export interface BackgroundActionRunOptions {
  /**
   * Additional runtime context data (ex. auth context data).
   */
  context?: ActionContext;

  /**
   * Additional span attributes to apply to OT spans.
   */
  telemetryLabels?: Record<string, string>;
}

/**
 * Options (side channel) data to pass to the model.
 */
export interface BackgroundActionFnArg<S> {
  /**
   * Additional runtime context data (ex. auth context data).
   */
  context?: ActionContext;

  /**
   * Trace context containing trace and span IDs.
   */
  trace: {
    traceId: string;
    spanId: string;
  };
}

/**
 * Action factory params.
 */
export type BackgroundActionParams<
  I extends z.ZodTypeAny,
  O extends z.ZodTypeAny,
  S extends z.ZodTypeAny = z.ZodTypeAny,
> = {
  name: string;
  start: (
    input: z.infer<I>,
    options: BackgroundActionFnArg<z.infer<S>>
  ) => Promise<Operation<z.infer<O>>>;
  check: (input: Operation<z.infer<O>>) => Promise<Operation<z.infer<O>>>;
  cancel?: (input: Operation<z.infer<O>>) => Promise<Operation<z.infer<O>>>;
  actionType: ActionType;

  description?: string;
  inputSchema?: I;
  inputJsonSchema?: JSONSchema7;
  outputSchema?: O;
  outputJsonSchema?: JSONSchema7;
  metadata?: Record<string, any>;
  use?: Middleware<z.infer<I>, z.infer<O>>[];
  streamSchema?: S;
};

/**
 * Defines an action with the given config and registers it in the registry.
 */
export function defineBackgroundAction<
  I extends z.ZodTypeAny,
  O extends z.ZodTypeAny,
  S extends z.ZodTypeAny = z.ZodTypeAny,
>(
  registry: Registry,
  config: BackgroundActionParams<I, O, S>
): BackgroundAction<I, O> {
  const startAction = defineAction(
    registry,
    {
      actionType: config.actionType,
      name: config.name,
      description: config.description,
      inputSchema: config.inputSchema,
      inputJsonSchema: config.inputJsonSchema,
      outputSchema: OperationSchema,
      metadata: {
        ...config.metadata,
        outputSchema: toJsonSchema({
          schema: config.outputSchema,
          jsonSchema: config.outputJsonSchema,
        }),
      },
      use: config.use,
    },
    async (input, options) => {
      const operation = await config.start(input, options);
      operation.action = `/${config.actionType}/${config.name}`;
      return operation;
    }
  );
  const checkAction = defineAction(
    registry,
    {
      actionType: 'check-operation',
      name: `${config.name}/check`,
      description: config.description,
      inputSchema: OperationSchema,
      inputJsonSchema: config.inputJsonSchema,
      outputSchema: OperationSchema,
      metadata: {
        ...config.metadata,
        outputSchema: toJsonSchema({
          schema: config.outputSchema,
          jsonSchema: config.outputJsonSchema,
        }),
      },
    },
    async (input) => {
      const operation = await config.check(input);
      operation.action = `/${config.actionType}/${config.name}`;
      return operation;
    }
  );
  let cancelAction:
    | Action<typeof OperationSchema, typeof OperationSchema>
    | undefined = undefined;
  if (config.cancel) {
    cancelAction = defineAction(
      registry,
      {
        actionType: 'cancel-operation',
        name: `${config.name}/cancel`,
        description: config.description,
        inputSchema: OperationSchema,
        inputJsonSchema: config.inputJsonSchema,
        outputSchema: OperationSchema,
        metadata: {
          ...config.metadata,
          outputSchema: toJsonSchema({
            schema: config.outputSchema,
            jsonSchema: config.outputJsonSchema,
          }),
        },
      },
      async (input) => {
        if (!config.cancel) {
          throw new GenkitError({
            status: 'UNAVAILABLE',
            message: `${config.name} does not support cancellation.`,
          });
        }
        const operation = await config.cancel(input);
        operation.action = `/${config.actionType}/${config.name}`;
        return operation;
      }
    );
  }

  return new BackgroundActionImpl(startAction, checkAction, cancelAction);
}

File: go/core/background_action.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package core

import (
	"context"

	"github.com/firebase/genkit/go/core/api"
)

// StartOpFunc starts a background operation.
type StartOpFunc[In, Out any] = func(ctx context.Context, input In) (*Operation[Out], error)

// CheckOpFunc checks the status of a background operation.
type CheckOpFunc[Out any] = func(ctx context.Context, op *Operation[Out]) (*Operation[Out], error)

// CancelOpFunc cancels a background operation.
type CancelOpFunc[Out any] = func(ctx context.Context, op *Operation[Out]) (*Operation[Out], error)

// Operation represents a long-running operation started by a background action.
type Operation[Out any] struct {
	Action   string         // Key of the action that created this operation.
	ID       string         // ID of the operation.
	Done     bool           // Whether the operation is complete.
	Output   Out            // Result when done.
	Error    error          // Error if the operation failed.
	Metadata map[string]any // Additional metadata.
}

// BackgroundActionDef is a background action that can be used to start, check, and cancel background operations.
//
// For internal use only.
type BackgroundActionDef[In, Out any] struct {
	*ActionDef[In, *Operation[Out], struct{}]

	check  *ActionDef[*Operation[Out], *Operation[Out], struct{}] // Sub-action that checks the status of a background operation.
	cancel *ActionDef[*Operation[Out], *Operation[Out], struct{}] // Sub-action that cancels a background operation.
}

// Start starts a background operation.
func (b *BackgroundActionDef[In, Out]) Start(ctx context.Context, input In) (*Operation[Out], error) {
	return b.Run(ctx, input, nil)
}

// Check checks the status of a background operation.
func (b *BackgroundActionDef[In, Out]) Check(ctx context.Context, op *Operation[Out]) (*Operation[Out], error) {
	return b.check.Run(ctx, op, nil)
}

// Cancel attempts to cancel a background operation. It returns an error if the background action does not support cancellation.
func (b *BackgroundActionDef[In, Out]) Cancel(ctx context.Context, op *Operation[Out]) (*Operation[Out], error) {
	if !b.SupportsCancel() {
		return nil, NewError(UNAVAILABLE, "model %q does not support canceling operations", b.Name())
	}

	return b.cancel.Run(ctx, op, nil)
}

// SupportsCancel returns whether the background action supports cancellation.
func (b *BackgroundActionDef[In, Out]) SupportsCancel() bool {
	return b.cancel != nil
}

// Register registers the model with the given registry.
func (b *BackgroundActionDef[In, Out]) Register(r api.Registry) {
	b.ActionDef.Register(r)
	b.check.Register(r)
	if b.cancel != nil {
		b.cancel.Register(r)
	}
}

// DefineBackgroundAction creates and registers a background action with three component actions
func DefineBackgroundAction[In, Out any](
	r api.Registry,
	name string,
	atype api.ActionType,
	metadata map[string]any,
	startFn StartOpFunc[In, Out],
	checkFn CheckOpFunc[Out],
	cancelFn CancelOpFunc[Out],
) *BackgroundActionDef[In, Out] {
	a := NewBackgroundAction(name, atype, metadata, startFn, checkFn, cancelFn)
	a.Register(r)
	return a
}

// NewBackgroundAction creates a new background action without registering it.
func NewBackgroundAction[In, Out any](
	name string,
	atype api.ActionType,
	metadata map[string]any,
	startFn StartOpFunc[In, Out],
	checkFn CheckOpFunc[Out],
	cancelFn CancelOpFunc[Out],
) *BackgroundActionDef[In, Out] {
	if name == "" {
		panic("core.NewBackgroundAction: name is required")
	}
	if startFn == nil {
		panic("core.NewBackgroundAction: startFn is required")
	}
	if checkFn == nil {
		panic("core.NewBackgroundAction: checkFn is required")
	}

	key := api.KeyFromName(atype, name)

	startAction := NewAction(name, atype, metadata, nil,
		func(ctx context.Context, input In) (*Operation[Out], error) {
			op, err := startFn(ctx, input)
			if err != nil {
				return nil, err
			}
			op.Action = key
			return op, nil
		})

	checkAction := NewAction(name, api.ActionTypeCheckOperation, metadata, nil,
		func(ctx context.Context, op *Operation[Out]) (*Operation[Out], error) {
			updatedOp, err := checkFn(ctx, op)
			if err != nil {
				return nil, err
			}
			updatedOp.Action = key
			return updatedOp, nil
		})

	var cancelAction *ActionDef[*Operation[Out], *Operation[Out], struct{}]
	if cancelFn != nil {
		cancelAction = NewAction(name, api.ActionTypeCancelOperation, metadata, nil,
			func(ctx context.Context, op *Operation[Out]) (*Operation[Out], error) {
				cancelledOp, err := cancelFn(ctx, op)
				if err != nil {
					return nil, err
				}
				cancelledOp.Action = key
				return cancelledOp, nil
			})
	}

	return &BackgroundActionDef[In, Out]{
		ActionDef: startAction,
		check:     checkAction,
		cancel:    cancelAction,
	}
}

// LookupBackgroundAction looks up a background action by key (which includes the action type, provider, and name).
func LookupBackgroundAction[In, Out any](r api.Registry, key string) *BackgroundActionDef[In, Out] {
	atype, provider, id := api.ParseKey(key)
	name := api.NewName(provider, id)

	startAction := ResolveActionFor[In, *Operation[Out], struct{}](r, atype, name)
	if startAction == nil {
		return nil
	}

	checkAction := ResolveActionFor[*Operation[Out], *Operation[Out], struct{}](r, api.ActionTypeCheckOperation, name)
	if checkAction == nil {
		return nil
	}

	cancelAction := ResolveActionFor[*Operation[Out], *Operation[Out], struct{}](r, api.ActionTypeCancelOperation, name)

	return &BackgroundActionDef[In, Out]{
		ActionDef: startAction,
		check:     checkAction,
		cancel:    cancelAction,
	}
}

// CheckOperation checks the status of a background operation by looking up the action and calling its Check method.
func CheckOperation[In, Out any](ctx context.Context, r api.Registry, op *Operation[Out]) (*Operation[Out], error) {
	if op == nil {
		return nil, NewError(INVALID_ARGUMENT, "core.CheckOperation: operation is nil")
	}

	if op.Action == "" {
		return nil, NewError(INVALID_ARGUMENT, "core.CheckOperation: operation is missing original request information")
	}

	m := LookupBackgroundAction[In, Out](r, op.Action)
	if m == nil {
		return nil, NewError(INVALID_ARGUMENT, "core.CheckOperation: failed to resolve background model %q from original request", op.Action)
	}

	return m.Check(ctx, op)
}
